test/config-luatest/downscaling_test.lua

```diff
@@ -0,0 +1,128 @@
+local t = require('luatest')
+local cbuilder = require('luatest.cbuilder')
+local cluster = require('test.config-luatest.cluster')
+
+local g = t.group()
+
+g.before_all(cluster.init)
+g.after_each(cluster.drop)
+g.after_all(cluster.clean)
+
+-- Verify that if an instance is removed from the declarative
+-- configuration, it is removed from upstreams in the box-level
+-- configuration (box.cfg.replication) after config:reload().
+--
+-- It effectively stops the data flow from the removed instance to
+-- others.
+g.test_size_3_to_2 = function(g)
+    local config = cbuilder:new()
+        :set_replicaset_option('replication.failover', 'manual')
+        :set_replicaset_option('leader', 'i-001')
+        :add_instance('i-001', {})
+        :add_instance('i-002', {})
+        :add_instance('i-003', {})
+        :config()
+
+    local cluster = cluster.new(g, config)
+    cluster:start()
+
+    -- Verify a test case prerequisite: all the instances have all
+    -- the instances in upstreams.
+    cluster:each(function(server)
+        server:exec(function()
+            local config = require('config')
+
+            t.assert_equals(box.cfg.replication, {
+                config:instance_uri('peer', {instance = 'i-001'}),
+                config:instance_uri('peer', {instance = 'i-002'}),
+                config:instance_uri('peer', {instance = 'i-003'}),
+            })
+        end)
+    end)
+
+    -- Remove i-003 from the configuration and write it to the
+    -- file.
+    local config_2 = cbuilder:new(config)
+        :set_replicaset_option('instances.i-003', nil)
+        :config()
+    cluster:sync(config_2)
+
+    -- Define a helper method.
+    cluster.each_except = function(self, exclusion, f)
+        self:each(function(server)
+            if server.alias ~= exclusion then
+                f(server)
+            end
+        end)
+    end
```

Comment:
I suppose it's better to change the sync method of the cluster not to iterate in each through the instances removed with `sync()`

Seems like, we should create a separate issue on it. What do you think?

Reply:
I agree that it would be convenient to have a support for temporary and permanent removing of instances from the cluster in the `cluster` testing helper. We should save the server objects into a separate field in the first case (to stop/drop them later), but we possibly should stop/drop the server object immediately in the second case. Also, a permanent removal means that we should go to a master instance and drop the corresponding `box.space._cluster` entry.

IOW, I'm in doubt about a proper API here, so postponed it. Maybe `:sync()` is not really what we want to have at the end and several more explicit methods are needed.

I prefer to return here back on demand, that's just our testing helper. I don't think that it needs an issue like a defect or a functionality in tarantool.

I hope it'll spin for a while in my head and I'll return back with some proposal.